### PR TITLE
iam: Improve ISO tag handling

### LINKED ISSUE #23286: IAM: Errors from tags in ISO regions
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Affected Resource(s)

* aws_iam_instance_profile

### Debug Output

```
error updating tags for IAM Instance Profile (<REDACTED>): Error tagging resource (<REDACTED>): InvalidAction: Unsupported operation.
```

### Expected Behavior

Able to update IAM Instance Profile in ISO region even if tags are not supported

### Actual Behavior

Per @caleb-devops [here](https://github.com/hashicorp/terraform-provider-aws/issues/22532#issuecomment-1035586195):

v3.74.0 did fix most of the errors relating to tagging, however I am receiving the following error during terraform apply for `aws_iam_instance_profile` resources:

```
error updating tags for IAM Instance Profile (<REDACTED>): Error tagging resource (<REDACTED>): InvalidAction: Unsupported operation.
```

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

ISO region

### References

* #22532

### PR DESCRIPTION
**Pull request by YakDriver**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #22532 
Closes #23286

Output from acceptance testing:

```console
% make testacc TESTS=TestAccIAMInstanceProfile_ PKG=iam
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/iam/... -v -count 1 -parallel 20 -run='TestAccIAMInstanceProfile_'  -timeout 180m
--- PASS: TestAccIAMInstanceProfile_Disappears_role (22.21s)
--- PASS: TestAccIAMInstanceProfile_disappears (22.49s)
--- PASS: TestAccIAMInstanceProfile_withoutRole (24.29s)
--- PASS: TestAccIAMInstanceProfile_basic (24.89s)
--- PASS: TestAccIAMInstanceProfile_namePrefix (24.90s)
--- PASS: TestAccIAMInstanceProfile_tags (46.92s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/iam	48.541s
% make testacc TESTS=TestAccIAMOpenIDConnectProvider_ PKG=iam
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/iam/... -v -count 1 -parallel 20 -run='TestAccIAMOpenIDConnectProvider_'  -timeout 180m
--- PASS: TestAccIAMOpenIDConnectProvider_disappears (12.61s)
--- PASS: TestAccIAMOpenIDConnectProvider_basic (27.61s)
--- PASS: TestAccIAMOpenIDConnectProvider_tags (37.40s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/iam	39.338s
% make testacc TESTS=TestAccIAMPolicy_ PKG=iam
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/iam/... -v -count 1 -parallel 20 -run='TestAccIAMPolicy_'  -timeout 180m
--- PASS: TestAccIAMPolicy_disappears (14.01s)
--- PASS: TestAccIAMPolicy_namePrefix (18.04s)
--- PASS: TestAccIAMPolicy_basic (18.04s)
--- PASS: TestAccIAMPolicy_description (18.08s)
--- PASS: TestAccIAMPolicy_path (18.09s)
--- PASS: TestAccIAMPolicy_policy (27.58s)
--- PASS: TestAccIAMPolicy_tags (36.86s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/iam	38.506s
% make testacc TESTS=TestAccIAMSAMLProvider PKG=iam
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/iam/... -v -count 1 -parallel 20 -run='TestAccIAMSAMLProvider'  -timeout 180m
--- PASS: TestAccIAMSAMLProvider_disappears (13.93s)
--- PASS: TestAccIAMSAMLProvider_basic (29.92s)
--- PASS: TestAccIAMSAMLProvider_tags (40.84s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/iam	42.488s
% make testacc TESTS=TestAccIAMServerCertificate PKG=iam
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/iam/... -v -count 1 -parallel 20 -run='TestAccIAMServerCertificate'  -timeout 180m
--- PASS: TestAccIAMServerCertificateDataSource_matchNamePrefix (4.66s)
--- PASS: TestAccIAMServerCertificate_disappears (17.80s)
--- PASS: TestAccIAMServerCertificate_Name_prefix (20.35s)
--- PASS: TestAccIAMServerCertificateDataSource_basic (20.68s)
--- PASS: TestAccIAMServerCertificateDataSource_path (20.77s)
--- PASS: TestAccIAMServerCertificate_path (22.88s)
--- PASS: TestAccIAMServerCertificate_basic (22.93s)
--- PASS: TestAccIAMServerCertificate_file (32.95s)
--- PASS: TestAccIAMServerCertificate_tags (46.10s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/iam	48.800s
% make testacc TESTS=TestAccIAMServiceLinkedRole PKG=iam     
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/iam/... -v -count 1 -parallel 20 -run='TestAccIAMServiceLinkedRole'  -timeout 180m
--- PASS: TestAccIAMServiceLinkedRole_disappears (25.11s)
--- PASS: TestAccIAMServiceLinkedRole_basic (30.96s)
--- PASS: TestAccIAMServiceLinkedRole_CustomSuffix_diffSuppressFunc (30.98s)
--- PASS: TestAccIAMServiceLinkedRole_customSuffix (31.02s)
--- PASS: TestAccIAMServiceLinkedRole_description (43.59s)
--- PASS: TestAccIAMServiceLinkedRole_tags (53.67s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/iam	55.516s
% make testacc TESTS=TestAccIAMVirtualMFADevice_ PKG=iam
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/iam/... -v -count 1 -parallel 20 -run='TestAccIAMVirtualMFADevice_'  -timeout 180m
--- PASS: TestAccIAMVirtualMFADevice_disappears (13.35s)
--- PASS: TestAccIAMVirtualMFADevice_basic (17.62s)
--- PASS: TestAccIAMVirtualMFADevice_tags (40.08s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/iam	42.001s
```

### CODE REVIEW

File: internal/service/iam/instance_profile.go [hunk 1]
```diff
@@ -117,6 +130,21 @@ func resourceInstanceProfileCreate(d *schema.ResourceData, meta interface{}) err
...
+		err := instanceProfileUpdateTags(conn, d.Id(), nil, tags)
+
+		// If default tags only, log and continue. Otherwise, error.
+		if v, ok := d.GetOk("tags"); (!ok || len(v.(map[string]interface{})) == 0) && verify.CheckISOErrorTagsUnsupported(err) {
+			log.Printf("[WARN] failed adding tags after create for IAM Instance Profile (%s): %s", d.Id(), err)
+			return resourceInstanceProfileUpdate(d, meta)
+		}
+
+		if err != nil {
+			return fmt.Errorf("error updating tags for IAM Instance Profile (%s): %w", d.Id(), err)
```
Comment (gdavison) [inline internal/service/iam/instance_profile.go:144]: Should this error message reference adding tags after creation? e.g.

```suggestion
			return fmt.Errorf("error adding tags after create for IAM Instance Profile (%s): %w", d.Id(), err)
```

File: internal/service/iam/saml_provider.go [hunk 2]
```diff
@@ -56,28 +56,55 @@ func ResourceSamlProvider() *schema.Resource {
...
+		err := samlProviderUpdateTags(conn, d.Id(), nil, tags)
+
+		// If default tags only, log and continue. Otherwise, error.
+		if v, ok := d.GetOk("tags"); (!ok || len(v.(map[string]interface{})) == 0) && verify.CheckISOErrorTagsUnsupported(err) {
+			log.Printf("[WARN] failed adding tags after create for IAM SAML Provider (%s): %s", d.Id(), err)
+			return resourceSAMLProviderRead(d, meta)
+		}
+
+		if err != nil {
+			return fmt.Errorf("error updating tags for IAM SAML Provider (%s): %w", d.Id(), err)
```
Comment (gdavison) [inline internal/service/iam/saml_provider.go:100]: Should the error match the warning?

```suggestion
			return fmt.Errorf("error adding tags after create for IAM SAML Provider (%s): %w", d.Id(), err)
```

File: internal/service/iam/service_linked_role.go [hunk 3]
```diff
@@ -112,7 +112,15 @@ func resourceServiceLinkedRoleCreate(d *schema.ResourceData, meta interface{}) e
...
+		err = roleUpdateTags(conn, roleName, nil, tags)
+
+		// If default tags only, log and continue. Otherwise, error.
+		if v, ok := d.GetOk("tags"); (!ok || len(v.(map[string]interface{})) == 0) && verify.CheckISOErrorTagsUnsupported(err) {
+			log.Printf("[WARN] failed adding tags after create for IAM Service Linked Role (%s): %s", d.Id(), err)
+			return resourceServiceLinkedRoleRead(d, meta)
+		}
+
+		if err != nil {
 			return fmt.Errorf("error updating IAM Service Linked Role (%s) tags: %w", d.Id(), err)
```
Comment (gdavison) [inline internal/service/iam/service_linked_role.go:124]: Here as well

```suggestion
			return fmt.Errorf("error adding tags after create for IAM Service Linked Role (%s): %w", d.Id(), err)
```

File: internal/service/iam/service_linked_role.go [hunk 4]
```diff
@@ -196,7 +204,15 @@ func resourceServiceLinkedRoleUpdate(d *schema.ResourceData, meta interface{}) e
 	if d.HasChange("tags_all") {
 		o, n := d.GetChange("tags_all")
 
-		if err := roleUpdateTags(conn, roleName, o, n); err != nil {
+		err := roleUpdateTags(conn, roleName, o, n)
+
+		// If default tags only, log and continue. Otherwise, error.
+		if v, ok := d.GetOk("tags"); (!ok || len(v.(map[string]interface{})) == 0) && verify.CheckISOErrorTagsUnsupported(err) {
+			log.Printf("[WARN] failed adding tags after create for IAM Service Linked Role (%s): %s", d.Id(), err)
```
Comment (gdavison) [inline internal/service/iam/service_linked_role.go:211]: This is in the update operation

```suggestion
			log.Printf("[WARN] failed updating tags for IAM Service Linked Role (%s): %s", d.Id(), err)
```

File: internal/service/iam/virtual_mfa_device.go [hunk 5]
```diff
@@ -78,24 +78,49 @@ func resourceVirtualMfaDeviceCreate(d *schema.ResourceData, meta interface{}) er
...
+		err := virtualMFAUpdateTags(conn, d.Id(), nil, tags)
+
+		// If default tags only, log and continue. Otherwise, error.
+		if v, ok := d.GetOk("tags"); (!ok || len(v.(map[string]interface{})) == 0) && verify.CheckISOErrorTagsUnsupported(err) {
+			log.Printf("[WARN] failed adding tags after create for IAM Virtual MFA Device (%s): %s", d.Id(), err)
+			return resourceVirtualMFADeviceRead(d, meta)
+		}
+
+		if err != nil {
+			return fmt.Errorf("error updating tags for IAM Virtual MFA Device (%s): %w", d.Id(), err)
```
Comment (gdavison) [inline internal/service/iam/virtual_mfa_device.go:111]:
```suggestion
			return fmt.Errorf("error adding tags after create for IAM Virtual MFA Device (%s): %w", d.Id(), err)
```

**Review by gdavison — COMMENTED** [review]
I've found a number of spots where the warning message and error message don't correspond when adding tags after creation

File: internal/service/iam/instance_profile.go (hunk 1 above)
Comment (YakDriver) [reply internal/service/iam/instance_profile.go:144]: Nice catch 👍

File: internal/service/iam/saml_provider.go (hunk 2 above)
Comment (YakDriver) [reply internal/service/iam/saml_provider.go]: Yikes

**Pull request merged**